File: Level 7 - System Design & Probabilistic DS/merkle_tree/core_problems.java
import java.util.*;

/**
 * Core Merkle Tree Problems in Java
 * Focuses on Merkle Proof (Audit Path) generation and verification.
 */
public class core_problems {

    static class Node {
        String hash;
        Node left, right;

        Node(String h) {
            this.hash = h;
        }
    }

    private List<Node> leaves;
    private String rootHash;

    public core_problems(List<String> data) {
        this.leaves = new ArrayList<>();
        for (String d : data)
            leaves.add(new Node(basic_operations.sha256(d)));
        this.rootHash = build(leaves).get(0).hash;
    }

    private List<Node> build(List<Node> currentLevel) {
        if (currentLevel.size() <= 1)
            return currentLevel;
        List<Node> parents = new ArrayList<>();
        for (int i = 0; i < currentLevel.size(); i += 2) {
            Node left = currentLevel.get(i);
            Node right = (i + 1 < currentLevel.size()) ? currentLevel.get(i + 1) : left;
            Node parent = new Node(basic_operations.sha256(left.hash + right.hash));
            parent.left = left;
            parent.right = right;
            parents.add(parent);
        }
        return build(parents);
    }

    // A proof consists of (sibling_hash, is_left_sibling) pairs
    static class ProofStep {
        String hash;
        boolean isLeft;

        ProofStep(String h, boolean l) {
            hash = h;
            isLeft = l;
        }
    }

    public List<ProofStep> getProof(int index) {
        List<ProofStep> proof = new ArrayList<>();
        int currIdx = index;
        List<Node> level = leaves;

        while (level.size() > 1) {
            int siblingIdx = (currIdx % 2 == 0) ? currIdx + 1 : currIdx - 1;
            if (siblingIdx < level.size()) {
                proof.add(new ProofStep(level.get(siblingIdx).hash, currIdx % 2 != 0));
            } else {
                // Odd length, sibling is self
                proof.add(new ProofStep(level.get(currIdx).hash, currIdx % 2 != 0));
            }
            currIdx /= 2;
            List<Node> nextLevel = new ArrayList<>();
            for (int i = 0; i < level.size(); i += 2) {
                Node left = level.get(i);
                Node right = (i + 1 < level.size()) ? level.get(i + 1) : left;
                nextLevel.add(new Node(basic_operations.sha256(left.hash + right.hash)));
            }
            level = nextLevel;
        }
        return proof;
    }

    public static boolean verify(String block, List<ProofStep> proof, String rootHash) {
        String currentHash = basic_operations.sha256(block);
        for (ProofStep step : proof) {
            if (step.isLeft) {
                currentHash = basic_operations.sha256(step.hash + currentHash);
            } else {
                currentHash = basic_operations.sha256(currentHash + step.hash);
            }
        }
        return currentHash.equals(rootHash);
    }

    public static void main(String[] args) {
        List<String> data = Arrays.asList("Tx1", "Tx2", "Tx3", "Tx4", "Tx5");
        core_problems mt = new core_problems(data);
        String root = mt.rootHash;

        System.out.println("Generating proof for 'Tx3' (index 2)...");
        List<ProofStep> proof = mt.getProof(2);

        System.out.println("Verifying Tx3: " + verify("Tx3", proof, root));
        System.out.println("Verifying Tx3 (Tampered): " + verify("Tx3_MOD", proof, root));
    }
}

File: Level 7 - System Design & Probabilistic DS/merkle_tree/core_problems.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <iomanip>
#include <sstream>

/**
 * Core Merkle Tree Problems in C++
 * Focuses on Proof (Audit Path) generation and verification.
 */

struct Node {
    std::string hash;
    Node *left, *right;
    Node(std::string h) : hash(h), left(nullptr), right(nullptr) {}
};

class MerkleTree {
private:
    std::string sha256(const std::string& input) {
        std::hash<std::string> hasher;
        size_t h = hasher(input);
        std::stringstream ss;
        ss << std::hex << std::setw(16) << std::setfill('0') << h;
        return ss.str();
    }

public:
    struct ProofStep {
        std::string hash;
        bool isLeft;
    };

    std::string rootHash;
    std::vector<std::string> leafHashes;

    MerkleTree(const std::vector<std::string>& data) {
        for (const auto& d : data) leafHashes.push_back(sha256(d));
        
        std::vector<std::string> level = leafHashes;
        while (level.size() > 1) {
            std::vector<std::string> next;
            for (size_t i = 0; i < level.size(); i += 2) {
                std::string left = level[i];
                std::string right = (i + 1 < level.size()) ? level[i+1] : left;
                next.push_back(sha256(left + right));
            }
            level = next;
        }
        rootHash = level.empty() ? "" : level[0];
    }

    std::vector<ProofStep> getProof(int index) {
        std::vector<ProofStep> proof;
        int curr = index;
        std::vector<std::string> level = leafHashes;

        while (level.size() > 1) {
            int sibling = (curr % 2 == 0) ? curr + 1 : curr - 1;
            if (sibling < (int)level.size()) {
                proof.push_back({level[sibling], curr % 2 != 0});
            } else {
                proof.push_back({level[curr], curr % 2 != 0});
            }
            curr /= 2;
            std::vector<std::string> next;
            for (size_t i = 0; i < level.size(); i += 2) {
                std::string l = level[i];
                std::string r = (i + 1 < level.size()) ? level[i+1] : l;
                next.push_back(sha256(l + r));
            }
            level = next;
        }
        return proof;
    }

    bool verify(const std::string& block, const std::vector<ProofStep>& proof) {
        std::string current = sha256(block);
        for (const auto& step : proof) {
            if (step.isLeft) current = sha256(step.hash + current);
            else current = sha256(current + step.hash);
        }
        return current == rootHash;
    }
};

int main() {
    std::vector<std::string> data = {"Tx1", "Tx2", "Tx3", "Tx4", "Tx5"};
    MerkleTree mt(data);

    std::cout << "Root: " << mt.rootHash << std::endl;
    std::cout << "Proof for Tx3..." << std::endl;
    auto proof = mt.getProof(2);

    std::cout << "Verify Tx3: " << (mt.verify("Tx3", proof) ? "Success" : "Failed") << std::endl;
    std::cout << "Verify Tx3 (Bad): " << (mt.verify("Tx3_BAD", proof) ? "Success" : "Failed") << std::endl;

    return 0;
}
